modules/custom/graphql_oauth/src/Entity/Server.php

```diff
@@ -0,0 +1,177 @@
+<?php
+
+namespace Drupal\graphql_oauth\Entity;
+
+use Drupal\Core\TypedData\Exception\MissingDataException;
+use Drupal\graphql\Entity\Server as OriginalServer;
+use Drupal\graphql\GraphQL\Execution\ResolveContext;
+use Drupal\graphql\GraphQL\ResolverRegistryInterface;
+use Drupal\simple_oauth\Authentication\TokenAuthUserInterface;
+use Drupal\simple_oauth\Oauth2ScopeInterface;
+use Drupal\simple_oauth\Plugin\Field\FieldType\Oauth2ScopeReferenceItemListInterface;
+use GraphQL\Error\UserError;
+use GraphQL\Executor\Values;
+use GraphQL\Type\Definition\Directive;
+use GraphQL\Type\Definition\ResolveInfo;
+use GraphQL\Type\Definition\WrappingType;
+
+/**
+ * Overrides the GraphQL configuration and request entry point.
+ */
+class Server extends OriginalServer {
+
+  /**
+   * {@inheritdoc}
+   */
+  protected function getFieldResolver(ResolverRegistryInterface $registry) {
+    $parent = parent::getFieldResolver($registry);
+    if (is_callable($parent)) {
+      $account_proxy = \Drupal::currentUser()->getAccount();
+      $token = NULL;
+      $token_scopes = [];
+      $directive_name = '';
+      if ($account_proxy instanceof TokenAuthUserInterface) {
+        $token = $account_proxy->getToken();
+        $scopes_field = $token->get('scopes');
+        assert($scopes_field instanceof Oauth2ScopeReferenceItemListInterface);
+        $token_scopes = array_map(function (Oauth2ScopeInterface $scope) {
+          return $scope->getName();
+        }, $scopes_field->getScopes());
+        $directive_name = !$token->get('auth_user_id')->isEmpty() ? 'allowUser' : 'allowBot';
+      }
+
+      return function ($value, $args, ResolveContext $context, ResolveInfo $info) use ($parent, $token, $directive_name, $token_scopes) {
+        $allow_directive_values = $this->getAllowDirectiveValues($info);
+        if (!empty($allow_directive_values)) {
+          // If token is not available; it has been removed or a different
+          // authentication is used. In this case we restrict access on
+          // types/fields that have the 'allow' directive.
+          if ($token === NULL) {
+            throw new MissingDataException('There is no OAuth2 token to work on.');
+          }
+          $this->checkAccess($allow_directive_values, $directive_name, $token_scopes);
+        }
+        return $parent($value, $args, $context, $info);
+      };
+    }
+    return $parent;
+  }
```

Comment:
On of the things that is slightly dangerous in the current set-up is that a future developer who does not understand the empty values of `$directive_name` and `$token` are linked may accidentally change the cases in which `$token` is changed while leaving `$directive_name` empty. That would cause `checkAccess` to be called with an invalid `$directive_name` value.

We can re-arrange the code by simply returning a different function depending on whether we have an account proxy or not. That can make our code simpler and more clearly communicate the intention to future developers.

```suggestion
  protected function getFieldResolver(ResolverRegistryInterface $registry) {
    $parent = parent::getFieldResolver($registry);
    if (is_callable($parent)) {
      $account_proxy = \Drupal::currentUser()->getAccount();

      if (!$account_proxy instanceof TokenAuthUserInterface) {
        // If the current user is not authenticated with OAuth then our field resolver 
        // simply denies access to any protected field because there is never a token.
        // That has the scope.
        return function ($value, $args, ResolveContext $context, ResolveInfo $info) use ($parent) {
          $allow_directive_values = $this->getAllowDirectiveValues($info);
          if (!empty($allow_directive_values)) {
              throw new MissingDataException('Token with valid scope was required but no token was presented.');
          }
          return $parent($value, $args, $context, $info);
          };
      }

      $token = $account_proxy->getToken();
      $scopes_field = $token->get('scopes');
      assert($scopes_field instanceof Oauth2ScopeReferenceItemListInterface);
      $token_scopes = array_map(function (Oauth2ScopeInterface $scope) {
          return $scope->getName();
      }, $scopes_field->getScopes());
      // `auth_user_id` is only set for tokens created for specific users.
      // Application ("bot") tokens are not linked to specific users but their user is 
      // determined by the user id configured for the consumer entity.
      $directive_name = !$token->get('auth_user_id')->isEmpty() ? 'allowUser' : 'allowBot';

      return function ($value, $args, ResolveContext $context, ResolveInfo $info) use ($parent, $token, $directive_name, $token_scopes) {
        $allow_directive_values = $this->getAllowDirectiveValues($info);
        if (!empty($allow_directive_values)) {
          $this->checkAccess($allow_directive_values, $directive_name, $token_scopes);
        }
        return $parent($value, $args, $context, $info);
      };
    }
    return $parent;
  }
```